### PR TITLE
Advertise operational MDNS parameters on device reboot

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
Device does not advertise operational parameters using the configured nodeid but rather hardcodes the nodeid used for testing.

 #### Summary of Changes
- Read the serialized nodeid from the adminpairing info that is serialized after commisioning (to be replaced once we have operational credentials)
- split out mdns operations to a separate file for easier inclusion/calling
- Enable multicast joining for IPv6 for esp32 as well - espressif sdk fork of LWIP does not seem to use the neighbour discover flag to determine if multicast is to be supported or not
- add a device event for 'got ip address' that is used to signal the mdns server to restart. This is because link local ipv6 addresses are not 'internet connected' yet they should trigger a listen and multicast join on the corresponding interfaces.

### CODE REVIEW

**Review by mspang — APPROVED** [review]

**Comment by mspang** [comment]
Upon provisioning esp32 with this patch, I'm seeing

```
E (874756) chip[IN]: UDP endpoint pool FULL
E (874766) chip[DIS]: Failed to start mDNS server: Inet Error 1000020 (0x000F4254): No more TCP endpoints
```

**Comment by andy31415** [comment]
> Upon provisioning esp32 with this patch, I'm seeing
> 
> ```
> E (874756) chip[IN]: UDP endpoint pool FULL
> E (874766) chip[DIS]: Failed to start mDNS server: Inet Error 1000020 (0x000F4254): No more TCP endpoints
> ```

I see this after GATT finishes - unsure why. Post reboot mDNS seems to work fine. I will debug it a bit.

**Comment by andy31415** [comment]
> > Upon provisioning esp32 with this patch, I'm seeing
> > ```
> > E (874756) chip[IN]: UDP endpoint pool FULL
> > E (874766) chip[DIS]: Failed to start mDNS server: Inet Error 1000020 (0x000F4254): No more TCP endpoints
> > ```
> 
> I see this after GATT finishes - unsure why. Post reboot mDNS seems to work fine. I will debug it a bit.

Should be fixed now: apparently the mdns server was doing udp->close which does not release the memory pool. Switched to udp->Free instead.

File: scripts/esp32_log_cat.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
+#!/usr/bin/env python
```
Comment (mspang) [inline scripts/esp32_log_cat.py:1]: Are you aware of

```
source examples/all-clusters-app/esp32/idf.sh      
export IDF_PATH=${HOME}/tools/esp-idf
idf make -C examples/all-clusters-app/esp32 monitor
```

?

If this is needed can you move it to a different PR?

**Review by mspang — APPROVED** [review]

File: scripts/esp32_log_cat.py (hunk 1 above)
Comment (andy31415) [reply scripts/esp32_log_cat.py]: Removed.

I am aware of monitor, however I generally build from vscode and monitor does not seem to allow as convenient filtering (like show warning and above only) nor does it show things as pretty.

File: src/lib/mdns/minimal/Server.cpp [hunk 2]
```diff
@@ -127,7 +127,7 @@ void ServerBase::Shutdown()
     {
         if (mEndpoints[i].udp != nullptr)
         {
-            mEndpoints[i].udp->Close();
+            mEndpoints[i].udp->Free();
```
Comment (msandstedt) [inline src/lib/mdns/minimal/Server.cpp:130]: Nice catch.

**Review by msandstedt — APPROVED** [review]

File: src/lib/mdns/minimal/Server.cpp (hunk 2 above)
Comment (andy31415) [reply src/lib/mdns/minimal/Server.cpp:130]: It was Michael - he noticed a error that was hidden for me through all other messages. Apparently after I made mDNS restart on every ip address get, this became more likely to happen.

**Comment by andy31415** [comment]
@woody-apple @bzbarsky-apple ?

**Review by woody-apple — APPROVED** [review]
